sql-plugin/src/main/311+-all/scala/org/apache/spark/sql/rapids/PCBSSchemaHelper.scala

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import java.util.concurrent.atomic.AtomicLong
+
+import scala.collection.mutable
+
+import com.nvidia.spark.rapids.shims.GpuTypeShims
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
+import org.apache.spark.sql.types.{ArrayType, AtomicType, ByteType, CalendarIntervalType, DataType, DataTypes, Decimal, DecimalType, IntegerType, LongType, MapType, NullType, StructField, StructType, UserDefinedType}
+
+object PCBSSchemaHelper {
+  private val mapping = new mutable.HashMap[DataType, DataType]()
```

Comment:
Is mutable HashMap thread safe?

Reply:
If this is a cache and we always convert thing the same way, then I guess it is okay, but it will grow unbounded and I am not sure I am okay with that.

Why do we not have an API that given an input data type produces the mapped output data type? That is all we need, I don't see why we need to have a cache for this and a map to go back and forth. It is not a one to one mapping CalendarInterval => Struct(Month, Day, microsec) but Struct(Month, Day, microsec) is not guaranteed to always map back to a CalendarInterval. So I don't see how this mapping is actually going to work correctly.

Reply:
Your earlier comment got me thinking and this is a big thing that was missed. The mapping will grow unbounded and I am not OK with it either. I was thinking about caching the map but your idea of having an API might work too. 

To answer your second question, I am always looking for the original dataType in the map so it should always work. e.g. when caching, map.put(calendarIntervalType => struct(m, d, ms)), but when looking up the mapping, I look for the original type because I have it in the selectedAttributes i.e. map.contains(calendarIntervalType) to know that this needs to be converted back.